### PR TITLE
Clean up the Challenge Protocol spec to be easier to edit by humans

### PR DESCRIPTION
**Pull request by mcy**

### CODE REVIEW

**Review by alistair23 — APPROVED** [review]

File: RoT/Protocol/Challenge_Protocol.md [hunk 1]
```diff
@@ -1,5688 +1,2854 @@
...
-
-
-    
-
-<p id="gdcalert6" ><span style="color: red; font-weight: bold">>>>>>  gd2md-html alert: inline image link here (to images/image6.png). Store image on your image server and adjust path/filename/extension if necessary. </span><br>(<a href="#">Back to top</a>)(<a href="#gdcalert7">Next alert</a>)<br><span style="color: red; font-weight: bold">>>>>> </span></p>
+All MCTP transactions are based on the SMBus Block Write bus protocol.  The
+following C-style struct shows an MCTP encapsulated message.
+
+```c
+struct MctpFrame {
```
Comment (chweimer) [inline RoT/Protocol/Challenge_Protocol.md]: We should move this to a table.  It doesn't have to look like the table in the original spec, but I think some kind of table representation would be more generally consumable.

File: RoT/Protocol/Challenge_Protocol.md [hunk 2]
```diff
@@ -1,5688 +1,2854 @@
...
+higher layer class of message being carried within the MCTP protocol.
 
 
+## Transport Layer Header
 
+The Management Component Transport Protocol (MCTP) Base Specification defines
+the MCTP packet header (refer to DSP0236 for field descriptions), reproduced
+here as a C-style struct.
+```c
+struct MctpPacket {
```
Comment (chweimer) [inline RoT/Protocol/Challenge_Protocol.md]: Similarly here, we should go back to some kind of table.

File: RoT/Protocol/Challenge_Protocol.md [hunk 3]
```diff
@@ -1,5688 +1,2854 @@
...
-   </td>
-   <td>Variable
-   </td>
-  </tr>
-</table>
+    Table 2 Vendor Defined Message
 
 
-Null (0) Source and Destination EIDs are typically supported, however AC-RoT devices that have multiple MCTP Endpoints may specify an EID value greater than 7 and less than 255.  The PA-RoT does not broadcast any MCTP messages.
+| -------------- | ---- | --- |
```
Comment (chweimer) [inline RoT/Protocol/Challenge_Protocol.md]: This doesn't seem to render well in github's viewer.  Is that a viewer-specific issue or there something else going on here.

File: RoT/Protocol/Challenge_Protocol.md [hunk 4]
```diff
@@ -1,5688 +1,2854 @@
...
-Authentication is the process used to establish trust in a specific device and prove integrity of the device firmware.  Once a device is authenticated, a secure session can optionally be established to provide confidentiality.  For some devices, a secure session can also be used to enable a cryptographic binding that can be used for additional security or to enable additional functionality.  
 
-A device only needs to support a single session from another endpoint.  In single session devices, the establishment of a new session will supersede and terminate the existing session to permit the new session.  The previous session will be terminated upon receiving an out of session (unencrypted) Get Digests request specifying a requested key exchange.  Get Digests requests received in session (encrypted) or without requesting key exchange will not cause the active session to terminate.  
+## Attestation Protocol Format
 
-## PA-RoT and AC-RoT Authentication
+Messages from a PA-RoT to an AC-RoT will have the following format, described
+as a C-style struct. This is embedded in the structure defined in Section 3.4.
+```c
+struct MessageBody {
```
Comment (chweimer) [inline RoT/Protocol/Challenge_Protocol.md]: Another instance where I think a table is preferred.

File: RoT/Protocol/Challenge_Protocol.md [hunk 5]
```diff
@@ -1,5688 +1,2854 @@
...
+| `Unseal Message`           |     | X    | 0x89 | O        | Begin an unsealing challenge.                       |
+| `Unseal Message Result`    |     |      | 0x8a | O        | Get unsealing status and result.                    |
+
+Command bytes `0xf0` through `0xff` will never be allocated by Cerberus, and may
+be treated as a "private use" area.
+
+A note on required messages:
+1. `Key Exchange` and `Session Sync` are required if any "Enc" messages are supported.
+2. `Get Attestation Data` is required if an Attestation Log is supported.
+3. If any update command (e.g., `Prepare`/`Update/`Activate` commands) are
```
Comment (chweimer) [inline RoT/Protocol/Challenge_Protocol.md]: Needs an extra back-tick after Update

**Review by chweimer — CHANGES_REQUESTED** [review]
This looks great.  Thanks for all the work on it.  Just a few minor comments, most of which we had already talked about.

File: RoT/Protocol/Challenge_Protocol.md (hunk 3 above)
Comment (mcy) [reply RoT/Protocol/Challenge_Protocol.md]: This was a formatting mistake on my part. Good catch.

File: RoT/Protocol/Challenge_Protocol.md [hunk 6]
```diff
@@ -1,5688 +1,2803 @@
...
+Once the certificate chain has been authenticated, the master will verify
+firmware integrity through the Challenge request.  The slave will generate a
+Challenge response signed with its Alias key which includes the measurement of
+device firmware.  The master can verify this measurement against a Component
+Firmware Manifest (CFM) or some other reference to confirm the firmware is
+valid.
+
+The authentication flow is as follows:
+
+1.  Master issues `GET_DIGTESTS` command for Slot 0
```
Comment (chweimer) [inline RoT/Protocol/Challenge_Protocol.md]: GET_DIGESTS

File: RoT/Protocol/Challenge_Protocol.md [hunk 7]
```diff
@@ -1,5688 +1,2803 @@
...
+2.  Slave responds with digests for the Alias certificate chain
+3.  For each certificate in the chain, Master checks if the certificate has been
+    cached.
+4.  If the certificate has not been cached, Master issues `GET_CERTIFICATE`
+    request.
+5.  Slave responds with the requestd certificate.
```
Comment (chweimer) [inline RoT/Protocol/Challenge_Protocol.md]: requested

File: RoT/Protocol/Challenge_Protocol.md [hunk 8]
```diff
@@ -1,5688 +1,2803 @@
...
 
-## PA-RoT and AC-RoT Authentication
+Messages from a PA-RoT to an AC-RoT will have the following format.
+This is embedded in the structure defined in Section 3.4.
+| Payload  | Description                              |
+|----------|------------------------------------------|
+| 0:0      | Integrity check flag.                    |
+| 7:1      | MCTP message type (always `0x7e`).       |
+| 23:8     | MCTP PCI Vendor ID (always `0x1414`).    |
+| 24:24    | Set if this is a request message.        |
```
Comment (chweimer) [inline RoT/Protocol/Challenge_Protocol.md]: This conversion lost some detail about this 'rq' bit.  It doesn't really distinguish between request/response messages.

From the original text:
Request Type
   </td>
   <td>This field indicates what type of request is contained in the message.  Messages defined in this specification shall have this bit set to 0.  Setting this bit to 1 provides a mechanism, aside from different vendor IDs, to support a device-specific command set.  Devices that don’t have any additional command support will return an error if this bit is 1.

File: RoT/Protocol/Challenge_Protocol.md [hunk 9]
```diff
@@ -1,5688 +1,2803 @@
...
+| 25:25    | Reserved (should be zero).               |
+| 26:26    | Set if encrypted (always, in this case). |
```
Comment (chweimer) [inline RoT/Protocol/Challenge_Protocol.md]: Messages are not always encrypted.

File: RoT/Protocol/Challenge_Protocol.md [hunk 10]
```diff
@@ -1,5688 +1,2803 @@
...
+vector in plaintext at the end of the message body.  The following table shows
+the body of an encrypted Cerberus message, with the encryption trailer. Note
+that offsets are given in bits, not bytes.
+
+| Payload   | Description                                                        |
+|-----------|--------------------------------------------------------------------|
+| 0:0       | Integrity check flag.                                              |
+| 7:1       | MCTP message type (always `0x7e`).                                 |
+| 23:8      | MCTP PCI Vendor ID (always `0x1414`).                              |
+| 24:24     | Set if this is a request message.                                  |
```
Comment (chweimer) [inline RoT/Protocol/Challenge_Protocol.md]: Same as the comment in the previous section.

**Comment by chweimer** [comment]
This is a lot of good work to reformat things for easier editing and diffs, but I did notice a couple places (the MCTP rq bit) where some information got lost.  I haven't read through the entire modified spec in detail to see if there are other places like this.  We have a couple options:
1. I read through in detail and see if I see anything else that is misrepresented.
2. We complete the PR as-is with the intended goal of reformatting the markdown.  I can then take an follow-up action to read through and edit an necessary to match the current spec.

Thoughts or preferences?

**Comment by mcy** [comment]
I think that depends on how much time you want to spend on it; some of the typos you pointed out are original, I think, and I plan to go through the whole thing and clean it up. (That said, I will fix what you've pointed out; I've definitely made mistakes copy/pasting tables around.)

Selfishly, I would prefer to get this in sooner, because then we can start doing further cleanups in parallel.

**Comment by chweimer** [comment]
From my perspective, either way I'm going to need to go through it.  It would probably be best to just complete this as-is with another round of cleanup as a follow-up.  It would probably also be easier to just directly edit anything that may be a mistake instead of going through PR comments.

**Comment by mcy** [comment]
Sounds good. I'll let you approve when you're comfortable with the state of the PR, and we can go from there.

**Comment by mcy** [comment]
Sounds good. I'll let you approve when you're comfortable with the state of the PR, and we can go from there.

**Comment by chweimer** [comment]
I think if we just clean up the few things that I've already noted, that would be good enough for a first pass.

File: RoT/Protocol/Challenge_Protocol.md (hunk 8 above)
Comment (mcy) [reply RoT/Protocol/Challenge_Protocol.md]: Oops. Shows I've had this wrong the whole time, haha.

**Comment by mcy** [comment]
Apparently I lied and I didn't have changes to push. Should all be addressed now, though!

**Review by chweimer — APPROVED** [review]

**Pull request merged**